### PR TITLE
tls: implement SPIFFE Certificate Validator for independent multiple trust domain support

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

This is the sequel to the previous refactoring PR: https://github.com/envoyproxy/envoy/pull/14757

Commit Message: tls: implement SPIFFE Certificate Validator for independent multiple trust domain support.
Additional Description: Adds the extension point for certificate validations, and its first implementation for SPIFFE multi trust domain support in a single listener or cluster. Resolves https://github.com/envoyproxy/envoy/issues/14614 and https://github.com/envoyproxy/envoy/issues/9284. 
Risk Level: low (only adding the  new extension point and one implementation for it)
Testing: unit tests and integration tests.
Docs Changes: 
Release Notes: tls: implement SPIFFE Certificate Validator for independent multiple trust domain support.

cc @lizan

### CODE REVIEW

File: api/envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto
Comment (lizan) [inline api/envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto:22]: add a comment what will be the key and what is is the value?

File: test/extensions/transport_sockets/tls/cert_validator/util.h
Comment (lizan) [inline test/extensions/transport_sockets/tls/cert_validator/util.h:21]: use CONSTRUCT_ON_FIRST_USE macro.

File: api/envoy/extensions/transport_sockets/tls/v3/common.proto
Comment (adisuissa) [inline api/envoy/extensions/transport_sockets/tls/v3/common.proto:379]: nit: Why is the field number 13 and not 12?

File: api/envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto
Comment (mathetake) [reply api/envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto:45]: added detailed documentation

File: generated_api_shadow/envoy/extensions/transport_sockets/tls/v4alpha/common.proto
Comment (evan2645) [inline generated_api_shadow/envoy/extensions/transport_sockets/tls/v4alpha/common.proto:380]: Sorry to wander in here, I have a quick question. Does this comment imply that I can't use `match_subject_alt_names` at the same time as `custom_validator_config`? Will the SPIFFE validator extension support name matching?

Wondering the same about `require_signed_certificate_timestamp` and `allow_expired_certificate`, both of which also seem like they'd still be desirable when using SPIFFE validation

Comment (evan2645) [reply generated_api_shadow/envoy/extensions/transport_sockets/tls/v4alpha/common.proto:380]: Related question: will a combined validation context work as expected if a custom validator config is set in one of them? i.e. can I set the name matching in a default context and then additionally configure the SPIFFE validator?

Comment (mathetake) [reply generated_api_shadow/envoy/extensions/transport_sockets/tls/v4alpha/common.proto:380]: Hi @evan2645, thanks for dropping by!

In the current implementation, Envoy does not use any of existing fields in `CertificateValidationContext.*` when `CertificateValidationContext.custom_validator_config` is set. So setting `match_subject_alt_names`, `require_signed_certificate_timestamp` and `allow_expired_certificate` would have no effect at all.

As for `require_signed_certificate_timestamp` and `allow_expired_certificate`, it's easy to implement, but I would like to limit the scope of this *initial* PR. So I would love to work on adding support for these fields in other PRs. (So I would change this description here to, like: `If specified, the usage of all the values above depends on the custom validator`.

For `match_subject_alt_names`, I would like to ask you about what the expected behavior looks like? When the name match, then how should we choose the trust domain?

**Comment by mathetake** [comment]
the flaky verify example is of websocket and nothing to do with this PR.  It seems like the same thing happens in other PRs. So I am making this ready for review.

File: generated_api_shadow/envoy/extensions/transport_sockets/tls/v4alpha/common.proto
Comment (evan2645) [reply generated_api_shadow/envoy/extensions/transport_sockets/tls/v4alpha/common.proto]: > Hi @evan2645, thanks for dropping by!

Thanks for the warm welcome :)

> it's easy to implement, but I would like to limit the scope of this initial PR.

Oh sure! Sorry, I didn't mean to imply that these things must be addressed right now. I just wanted to call attention to it, and make sure that it is at least being considered in the grand scheme of things. I'd probably call support for `require_signed_certificate_timestamp` and `allow_expired_certificate` as nice to have. The feature provided by `match_subject_alt_names` configurable is much more important.

> So I would change this description here to, like: If specified, the usage of all the values above depends on the custom validator.

That sounds pretty good, especially if it means we'll be able to use the existing fields/values rather than having to redefine them on the custom validator portion of the proto

> ~~For match_subject_alt_names, I would like to ask you about what the expected behavior looks like? When the name match, then how should we choose the trust domain?~~ I got it. Choose the trust bundle based on the given trust domain first, and then verify the SAN name matching, and so maybe you want to have SAN matching conditions per trust bundle, right?

It could be that I want to allow multiple SPIFFE IDs that are in different trust domains. To support this, what we'll probably want to do is:
1. Extract peer SPIFFE ID
1. Choose correct bundle based on trust domain
1. Validate peer SVID using the selected bundle
1. Compare the validated SPIFFE ID to the allowed names from `match_subject_alt_names`

I don't think that we want to tie the matching to a particular bundle.

> Related question: will a combined validation context work as expected if a custom validator config is set in one of them? i.e. can I set the name matching in a default context and then additionally configure the SPIFFE validator?

I _think_ this would work as expected if the custom validator is able to consume the existing fields/values from the validation context proto?

The important part here is that the SPIFFE identity and bundles may be supplied by one system, and the authorization policy (e.g. `match_subject_alt_names`) from a different system. Since the config for authentication and authorization are mixed into the same proto, the only way we can (currently) solve for this is through the use of a combined validation context in which `match_subject_alt_names` is provided by one and `custom_validator_config` is provided by the other. If there is a better way to accomplish this, I'd love to hear about it

Comment (mathetake) [reply generated_api_shadow/envoy/extensions/transport_sockets/tls/v4alpha/common.proto]: > I think this would work as expected if the custom validator is able to consume the existing fields/values from the validation context proto?

> The important part here is that the SPIFFE identity and bundles may be supplied by one system, and the authorization policy (e.g. match_subject_alt_names) from a different system. Since the config for authentication and authorization are mixed into the same proto, the only way we can (currently) solve for this is through the use of a combined validation context in which match_subject_alt_names is provided by one and custom_validator_config is provided by the other. If there is a better way to accomplish this, I'd love to hear about it

OK, that makes sense. Thanks for the detailed description :+1:

File: api/envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto
Comment (lizan) [inline api/envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto:45]: Will this have more field than just a certificate chain? It might makes more sense to define a message TrustBundle and use it here.

File: include/envoy/ssl/BUILD
Comment (lizan) [inline include/envoy/ssl/BUILD:71]: delete

File: generated_api_shadow/envoy/extensions/transport_sockets/tls/v4alpha/common.proto
Comment (evan2645) [reply generated_api_shadow/envoy/extensions/transport_sockets/tls/v4alpha/common.proto]: Thanks for your work on this @mathetake please let me know if there's anything else I can do to help ❤️

File: source/extensions/transport_sockets/tls/cert_validator/spiffe_validator.cc
Comment (lizan) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe_validator.cc:210]: You might want to do manual parse instead of std::regex, this is not efficient.

File: test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc
Comment (lizan) [inline test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc:359]: where is this released? you just need a `envoy::config::core::v3::TypedExtensionConfig` instead of a pointer.

File: source/extensions/transport_sockets/tls/cert_validator/spiffe_validator.h
Comment (lizan) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe_validator.h:1]: Structurally this validator should be in a separate directory, and build with `envoy_cc_extension`.

File: test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc
Comment (mathetake) [reply test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc:359]: the pointer is passed to `validationContext`, and owned and freed there. That's why I created a raw pointer here to avoid double free.

**Review by lizan — COMMENTED** [review]
This looks good overall, but can you split the default validator refactoring to a different PR?

**Comment by mathetake** [comment]
@lizan could you take another look?

**Comment by lizan** [comment]
/lgtm api

Where is the reverted part PR, or do you want to keep default validator non-extension? 

This PR itself looks good to me. @ggreenway for non-Tetrands review.

**Comment by mathetake** [comment]
> do you want to keep default validator non-extension?

I'm now in favor of keeping default validator non-extension, but I don't have any concrete reason for that. So I would like to have you and @ggreenway's opinion. WDYT?

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (ggreenway) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:223]: `auto pos = san.find('/', prefix.size())); if (pos != std::string::npos) { return san.substr(prefix.size(), pos); }`

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (ggreenway) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:235]: I think the interface is adequate for this. The correct behavior is to loop over the certs, and return the minimum value of all of them.

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (ggreenway) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:256]: `CertValidatorValues::SPIFFE`

File: include/envoy/ssl/context_config.h
Comment (ggreenway) [inline include/envoy/ssl/context_config.h:53]: Why did this become non-const?

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (ggreenway) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:185]: Capitalize, period at end.

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall.

Please go through and fix comments to start with a capital letter and end with a period (except the 1-2 word comments describing a test case).

**Comment by ggreenway** [comment]
> > do you want to keep default validator non-extension?
> 
> I'm now in favor of keeping default validator non-extension, but I don't have any concrete reason for that. So I would like to have you and @ggreenway's opinion. WDYT?

What's the context of this question? Is it whether we should apply the default-validator in addition to the custom-validator?

File: include/envoy/ssl/context_config.h
Comment (mathetake) [reply include/envoy/ssl/context_config.h:53]: This is because I added `Api::Api& api()` to `CertificateValidationContextConfig` which cannot be const if I understand correctly. API object is necessary for SPIFFE validator to retrieve trust bundles from DataSource. If there's a workaround, I would like to adopt. thoughts?

**Comment by mathetake** [comment]
@ggreenway I mean whether we should use "envoy_cc_extension" for default validation to be aligned with SPIFFE validator as I did in this PR. In anyway, that should be done in another PR and doesn't affect this PR.

**Comment by mathetake** [comment]
coverage failure is irrelevant to this PR (same happens to other PRs): `Code coverage for source/server is lower than limit of 94.5 (94.4)`.

/retest

File: include/envoy/ssl/context_config.h
Comment (mathetake) [reply include/envoy/ssl/context_config.h]: https://github.com/envoyproxy/envoy/pull/14884/commits/63e4901a74abcef1910d91f410180128de4dc6e5 fixed!

**Comment by mathetake** [comment]
@ggreenway I think all the comments are addressed. PTAL 👍

**Review by ggreenway — DISMISSED** [review]
Code LGTM, but I am not very knowledable about spiffe. Someone else needs to sign-off on the spiffe validation logic specifically.

**Comment by azdagron** [comment]
This is super exciting! I'm a maintainer on the SPIFFE and SPIRE projects and am happy to look at the validation logic.

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (azdagron) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:76]: nit: I assume this added ref count is for storing in `ca_certs_`? I'm not familiar with Envoy policies on exception handling due to out-of-memory situations, but if the ref is added after the call to push_back() then `item-x509` would still be properly freed if an exception were to occur with the `push_back()`.

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (azdagron) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:107]: I'd like to understand why we enforce a unique set based on the subject name of the CA certificates? There is no constraint in the SPIFFE specification that X.509 authorities in a trust bundle have unique subject names. In fact, in practice, SPIRE server CAs part of the same deployment often have the same subject name for the CA certificates. Is there a particular situation we're trying to avoid here?

**Review by azdagron — COMMENTED** [review]
I'll pick this up again tomorrow but I had two initial comments. Again, super excited for this to land!

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (mathetake) [reply source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:76]: we've already reserve the memory for map before entering this for-loop, so no relocation happens  in push_back which is exception-free. So I think this is fine.

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (azdagron) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:183]: I think this is only checking the first of any SAN type, but from the spec:
>  An X.509 SVID MUST contain exactly one URI SAN. It MAY contain any number of other SAN fields, including DNS SANs.

I think we want something like the following:
```suggestion
  for (const GENERAL_NAME* general_name : san_names.get()) {
     if (generate_name->type != GEN_URI) {
         continue;
     }
    const std::string san = Utility::generalNameAsString(general_name);
```

**Review by azdagron — COMMENTED** [review]
I've finished my review. This is looking great. I appreciate the keen eye towards enforcing the specification.

As pointed out, there is one discrepancy in the SAN handling. I also had an earlier comment about some worrisome behavior that will filter out X.509 authorities in the trust bundle that happen to share a subject name. Outside of that I give a hearty thumbs up.

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (mathetake) [reply source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc]: Thanks! resolved in https://github.com/envoyproxy/envoy/pull/14884/commits/e6b2c0fdca1e411e814d097457f958b8dda5b84c

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (mathetake) [reply source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc]: resolved in  https://github.com/envoyproxy/envoy/pull/14884/commits/cdfb3ced2b5125432a40c56697164eedab8f3d15

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (mathetake) [reply source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:107]: This logic is what I borrowed from the existing certificate validation logic in Envoy, and not sure why it does deduplication on SAN. It's been 3+ years since [the PR was merged](https://github.com/envoyproxy/envoy/pull/2248). I guess this is only for `SSL_CTX_set_client_CA_list` and the openssl doc `The CAs listed do not become trusted (list only contains the names, not the complete certificates)`. 

cc @PiotrSikora since you are the author of the PR and an expert in this field, and could you give me the reason why the list for `SSL_CTX_set_client_CA_list` is deduplicated on SAN names?

Comment (PiotrSikora) [reply source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:107]: `SSL_CTX_set_client_CA_list` sets only a list of names, not a list of trusted CAs.

This function (well, not this one exactly, but the one this is a copy of) is a non-file based re-implementation of [`SSL_load_client_CA_file`](https://github.com/google/boringssl/blob/0a86cbff452f3c1807738968a88332416279e3dc/ssl/ssl_file.cc#L167)  which contains the same logic.

Comment (mathetake) [reply source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:107]: Thanks, that makes sense. I believe this is fine as it is, so I'm marking this comment as resolved.

**Comment by mathetake** [comment]
@lizan I believe we are good to go now. Could you take another look? Thanks!

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (azdagron) [reply source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:107]: I see now. Thanks for the answer!

**Comment by lizan** [comment]
@mathetake two last things
- add an entry in CODEOWNERS the new extension
- add a note in version history

**Comment by mathetake** [comment]
@lizan done!

File: docs/root/version_history/current.rst
Comment (lizan) [inline docs/root/version_history/current.rst:119]: This is not only listener but cluster as well?

Comment (mathetake) [reply docs/root/version_history/current.rst]: thanks

**Review by lizan — APPROVED** [review]

File: api/envoy/extensions/transport_sockets/tls/v3/common.proto
Comment (mattklein123) [inline api/envoy/extensions/transport_sockets/tls/v3/common.proto:385]: cc @phlax @mathetake this one needs to be updated for the new extension documentation mechanism as a follow up.

File: docs/root/version_history/current.rst
Comment (mattklein123) [inline docs/root/version_history/current.rst:126]: @mathetake this should also be documented here: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/security somewhere. Can you do this in a follow up while also fixing to use the automated extension documentation system? @phlax can help you out. Thank you!

File: api/envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto
Comment (mattklein123) [inline api/envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto:24]: Please update this to use validating-code-block

File: test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc
Comment (antoniovicente) [inline test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc:356]: The inclusion of SPIFFE tests in ssl_integration_test breaks builds for folks that want to include the TLS extension but not the SPIFFE cert verifier extension.  Would it be possible to move these tests to a separate test target?

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (antoniovicente) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:67]: Coverage report indicates there is no coverage for 4 throws in this file, plus several other error handling branches.

https://storage.googleapis.com/envoy-pr/684420f/coverage/source/extensions/transport_sockets/tls/cert_validator/default_validator.cc.gcov.html

File: test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc
Comment (mathetake) [reply test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc:356]: Thanks, I did in https://github.com/envoyproxy/envoy/pull/15324

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (mathetake) [reply source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:67]: That is for default_validator, not for SPIFFE one. I see these are uncovered quite a while.

**Pull request merged**
